src/Derivative.cpp

```diff
@@ -557,26 +557,39 @@ void ReverseAccumulationVisitor::propagate_adjoints(
         // non overwriting scans, we delay the boundary condition
         // setup since the gradients depend on itself.
         auto add_boundary_condition = [&](const FuncKey &func_key) {
+            // Add an update stage to adjoint_func to set everything out of bounds zero.
+            // We don't use BoundaryConditions::constant_exterior since it introduces 
+            // two extra Funcs, which
+            // 1) makes generated derivatives code less readable, and
+            // 2) makes (auto-)scheduling more difficult since there are more Funcs to consider.
             Func &adjoint_func = adjoint_funcs[func_key];
-            const Box &bounds = func_bounds[func.name()];
-            // Save a pointer to the unbounded def. Useful for scheduling
-            FuncKey unbounded_func_key{func.name() + "_unbounded", func_key.second};
-            adjoint_funcs[unbounded_func_key] = adjoint_func;
+            vector<pair<Expr, Expr>> bounds = box_to_vector(func_bounds[func.name()]);
+            vector<Var> args = adjoint_func.args();
+            Expr out_of_bounds = cast<bool>(false);
+            for (size_t i = 0; i < bounds.size(); i++) {
+                Var arg_var = args[i];
+                Expr min = bounds[i].first;
+                Expr extent = bounds[i].second;
+                internal_assert(min.defined() && extent.defined()) <<
+                    "Unbounded adjoint function detected.";
+
+                out_of_bounds = (out_of_bounds ||
+                                 arg_var < min ||
+                                 arg_var >= min + extent);
+            }
             if (adjoint_func.values().size() == 1) {
-                Type type = adjoint_func.values()[0].type();
-                internal_assert(adjoint_func.function().output_types()[0] ==
-                                adjoint_func.values()[0].type());
-                Func f = BoundaryConditions::constant_exterior(
-                    adjoint_func, make_const(type, 0.0), box_to_vector(bounds));
-                adjoint_func = f;
-
+                Expr zero = make_const(adjoint_func.value().type(), 0.0);
```

Comment:
make_zero(adjoint_func.value().type())

Reply:
There are too many make_const in Derivative.cpp. Fixed in https://github.com/halide/Halide/pull/4325